### PR TITLE
Hash based figure tests with pytest-mpl

### PR DESCRIPTION
**Pull request by Cadair**
This is a work in progress port to using pytest-mpl with new hash based functionality implemented in https://github.com/matplotlib/pytest-mpl/pull/101

TODO:
- [x] Await merge and release upstream
- [x] Fix the baseline image updates after merge.

### CODE REVIEW

File: tox.ini
Comment (Cadair) [inline tox.ini:91]: If we ever needed a different set of figures per-branch we would have to manually change this URL.

**Comment by Cadair** [comment]
We can't merge this until the pytest-mpl PRs land and a release happens, but it would be good to get a review from this side so I can adjust in pytest-mpl as needed.

File: sunpy/map/tests/test_plotting.py
Comment (nabobalis) [inline sunpy/map/tests/test_plotting.py:90]: This changes just to break the CI?

File: sunpy/tests/helpers.py
Comment (nabobalis) [inline sunpy/tests/helpers.py:88]: Do we need to keep this?

File: tox.ini
Comment (nabobalis) [reply tox.ini:91]: This is a good question, so far we have not needed this but I wonder if that will always be the case.

File: setup.cfg
Comment (dstansby) [inline setup.cfg:130]: Why is this warning being thrown?

File: sunpy/map/tests/test_plotting.py
Comment (dstansby) [inline sunpy/map/tests/test_plotting.py:90]: Why these changes?

File: sunpy/tests/helpers.py
Comment (dstansby) [inline sunpy/tests/helpers.py:88]: Is the plan to leave this here forever?

File: sunpy/tests/helpers.py
Comment (dstansby) [inline sunpy/tests/helpers.py:162]: Noice, did you put the HTML comparison stuff in pytest-mpl too?

**Review by dstansby — COMMENTED** [review]
Looks good overall; I've left a few comments.

File: sunpy/map/tests/test_plotting.py
Comment (Cadair) [reply sunpy/map/tests/test_plotting.py:90]: yep

File: sunpy/tests/helpers.py
Comment (Cadair) [reply sunpy/tests/helpers.py:162]: yep

File: setup.cfg
Comment (nabobalis) [reply setup.cfg:130]: I thought this warning was thrown when you set MPL to use agg. I have seen this locally at times with the figure tests.

File: sunpy/tests/helpers.py
Comment (Cadair) [reply sunpy/tests/helpers.py:88]: Maybe we don't need it at all. The reason to keep it was to keep `-m "figure"` working

File: sunpy/map/tests/test_plotting.py
Comment (Cadair) [reply sunpy/map/tests/test_plotting.py:90]: Just to break the tests, so I could see if the HTML worked.

File: setup.cfg
Comment (Cadair) [reply setup.cfg:130]: pytest-mpl dynamically switches the backend, but I don't understand how we avoided this in our own code before this PR. As I thought we set the backend to Agg anyway?

File: sunpy/tests/helpers.py
Comment (nabobalis) [reply sunpy/tests/helpers.py:88]: What would it be without this? `-m mpl-figure`?

File: setup.cfg
Comment (nabobalis) [reply setup.cfg:130]: IN tox we do set it to AGG and I have seen this warning but it is a good question on how have we avoided this before.

File: sunpy/tests/helpers.py
Comment (Cadair) [reply sunpy/tests/helpers.py:88]: `-m mpl_image_compare` as done in tox.ini

File: setup.cfg
Comment (Cadair) [reply setup.cfg:130]: we set Agg in the conftest.py as well: https://github.com/sunpy/sunpy/blob/master/sunpy/conftest.py#L17-L22

File: sunpy/tests/helpers.py
Comment (nabobalis) [reply sunpy/tests/helpers.py:88]: That isn't so bad. I feel like we could remove "figure"

File: setup.cfg
Comment (nabobalis) [reply setup.cfg:130]: Do we need both?

Comment (Cadair) [reply setup.cfg:130]: :man_shrugging:

Comment (dstansby) [reply setup.cfg:130]: What's the stack trace when this warning is thrown? Maybe pytest-mpl needs to be setting it to agg somewhere?

File: sunpy/map/tests/test_plotting.py
Comment (dstansby) [reply sunpy/map/tests/test_plotting.py:90]: 👍 will leave this un-resolved so we don't forget to remove

File: sunpy/tests/helpers.py
Comment (dstansby) [reply sunpy/tests/helpers.py:88]: Yeah, I would say that there are probably so few people running the figure tests locally that changing this is fine.

File: setup.cfg
Comment (Cadair) [reply setup.cfg:130]: I think this is being raised because pytest-mpl is setting it to agg and we are calling .show inside of our peek methods.

Comment (Cadair) [reply setup.cfg:130]: I can't replicate this inside tox, but the tests should run outside it just as well. The full traceback is:

```
___________________________________________________________ test_heliographic_peek ____________________________________________________________

cls = <class 'matplotlib.backends.backend_agg._BackendAgg'>

    @classmethod
    def show(cls, *, block=None):
        """
        Show all figures.
    
        `show` blocks by calling `mainloop` if *block* is ``True``, or if it
        is ``None`` and we are neither in IPython's ``%pylab`` mode, nor in
        `interactive` mode.
        """
        managers = Gcf.get_all_fig_managers()
        if not managers:
            return
        for manager in managers:
            try:
>               manager.show()  # Emits a warning for non-interactive backend.

../../.virtualenvs/pytest-mpl/lib/python3.8/site-packages/matplotlib/backend_bases.py:3524: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <matplotlib.backend_bases.FigureManagerBase object at 0x7ff5ab60a1c0>

    def show(self):
        """
        For GUI backends, show the figure window and redraw.
        For non-GUI backends, raise an exception, unless running headless (i.e.
        on Linux with an unset DISPLAY); this exception is converted to a
        warning in `.Figure.show`.
        """
        # This should be overridden in GUI backends.
        if cbook._get_running_interactive_framework() != "headless":
>           raise NonGuiException(
                f"Matplotlib is currently using {get_backend()}, which is "
                f"a non-GUI backend, so cannot show the figure.")
E           matplotlib.backend_bases.NonGuiException: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.

../../.virtualenvs/pytest-mpl/lib/python3.8/site-packages/matplotlib/backend_bases.py:2728: NonGuiException

During handling of the above exception, another exception occurred:

heliographic_test_map = <sunpy.map.mapbase.GenericMap object at 0x7ff571b1b0a0>
SunPy Map
---------
Observatory:             
Instrument:              
Detector:               ...    -0.02115 ],
       [-0.030485, -0.030435, -0.030385, ..., -0.03059 , -0.030535,
        -0.030485]], dtype=float32)

    @figure_test
    def test_heliographic_peek(heliographic_test_map):
>       heliographic_test_map.peek()

sunpy/map/tests/test_plotting.py:179: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sunpy/visualization/visualization.py:22: in show_figure
    plt.show()
../../.virtualenvs/pytest-mpl/lib/python3.8/site-packages/matplotlib/pyplot.py:353: in show
    return _backend_mod.show(*args, **kwargs)
../../.virtualenvs/pytest-mpl/lib/python3.8/site-packages/matplotlib/backend_bases.py:3526: in show
    cbook._warn_external(str(exc))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

message = 'Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.', category = None

    def _warn_external(message, category=None):
        """
        `warnings.warn` wrapper that sets *stacklevel* to "outside Matplotlib".
    
        The original emitter of the warning can be obtained by patching this
        function back to `warnings.warn`, i.e. ``cbook._warn_external =
        warnings.warn`` (or ``functools.partial(warnings.warn, stacklevel=2)``,
        etc.).
        """
        frame = sys._getframe()
        for stacklevel in itertools.count(1):  # lgtm[py/unused-loop-variable]
            if frame is None:
                # when called in embedded context may hit frame is None
                break
            if not re.match(r"\A(matplotlib|mpl_toolkits)(\Z|\.(?!tests\.))",
                            # Work around sphinx-gallery not setting __name__.
                            frame.f_globals.get("__name__", "")):
                break
            frame = frame.f_back
>       warnings.warn(message, category, stacklevel)
E       UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.

../../.virtualenvs/pytest-mpl/lib/python3.8/site-packages/matplotlib/cbook/__init__.py:2108: UserWarning
```

Comment (dstansby) [reply setup.cfg:132]: 🤦 of course this is exactly what is expected, silly me for not realising. Could you drop a comment in above the warning ignore along the lines of ".peek() calls plt.show(), that raises a warning in the test suite because we use the agg backend"

**Comment by Cadair** [comment]
Ok I think this is now ready to go. The pytest-mpl release it out, and I am pretty sure I have sorted the upload to the figure repo in the circle ci config now as well.

**Review by nabobalis — APPROVED** [review]

File: .circleci/config.yml
Comment (dstansby) [inline .circleci/config.yml:74]: Can the depth of this clone be reduced (perhaps to 1)?

File: .circleci/config.yml
Comment (dstansby) [inline .circleci/config.yml:79]: This seems hacky, is there any other way to extract the hash library name (perhaps from the environment name)?

**Review by dstansby — CHANGES_REQUESTED** [review]
Seems good overall; could you add a removals changelog that states that `sunpy/tests/hash.py` has gone. Not documented and not public API so fine to delete the code, but you never know if someone out there is using it...

File: .circleci/config.yml
Comment (Cadair) [reply .circleci/config.yml:79]: Not that I can think of, which is why we ended up here. I don't want to put a manual map of tox name to lib versions in the circle ci config (by hard coding the hash name for each env var). Only by extracting the hash name from the code that generates it in the pytest decorator can we be sure to get it right. I think my regex is pretty restrictive, so it should only match the explicit print line in the tox.ini.

**Review by nabobalis — APPROVED** [review]
Assuming codecov is ok.

**Comment by dstansby** [comment]
Since this PR only affects tests, I'm going to remove the 2.1 milestone

**Comment by Cadair** [comment]
@dstansby are you ok with this as it is?

File: setup.cfg
Comment (dstansby) [inline setup.cfg:103]: Does this fix https://github.com/sunpy/sunpy/issues/4612?

**Review by dstansby — APPROVED** [review]

File: setup.cfg
Comment (Cadair) [reply setup.cfg:103]: I think it fixes some of the issues. I don't *think* it would fix the asdf issue tho.

**Pull request merged**
